RecoLocalCalo/HGCalRecProducers/plugins/alpaka/HGCalSoARecHitsProducer.cc

```diff
@@ -0,0 +1,216 @@
+#include "DataFormats/HGCalReco/interface/alpaka/HGCalSoARecHitsDeviceCollection.h"
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+
+  class HGCalSoARecHitsProducer : public stream::EDProducer<> {
+  public:
+    HGCalSoARecHitsProducer(edm::ParameterSet const& config)
+        : detector_(config.getParameter<std::string>("detector")),
+          initialized_(false),
+          isNose_(detector_ == "HFNose"),
+          maxNumberOfThickIndices_(config.getParameter<unsigned>("maxNumberOfThickIndices")),
+          fcPerEle_(config.getParameter<double>("fcPerEle")),
+          ecut_(config.getParameter<double>("ecut")),
+          fcPerMip_(config.getParameter<std::vector<double>>("fcPerMip")),
+          nonAgedNoises_(config.getParameter<std::vector<double>>("noises")),
+          dEdXweights_(config.getParameter<std::vector<double>>("dEdXweights")),
+          thicknessCorrection_(config.getParameter<std::vector<double>>("thicknessCorrection")),
+          caloGeomToken_(consumesCollector().esConsumes<CaloGeometry, CaloGeometryRecord>()),
+          hits_token_(consumes<HGCRecHitCollection>(config.getParameter<edm::InputTag>("recHits"))),
+          deviceToken_{produces()} {}
+
+    ~HGCalSoARecHitsProducer() override = default;
+
+    void produce(device::Event& iEvent, device::EventSetup const& iSetup) override {
+      edm::Handle<HGCRecHitCollection> hits_h;
+
+      edm::ESHandle<CaloGeometry> geom = iSetup.getHandle(caloGeomToken_);
+      rhtools_.setGeometry(*geom);
+      maxlayer_ = rhtools_.lastLayer(isNose_);
+
+      hits_h = iEvent.getHandle(hits_token_);
+      auto const& hits = *(hits_h.product());
+      computeThreshold();
+
+      // Count effective hits above threshold
+      uint32_t index = 0;
+      for (unsigned int i = 0; i < hits.size(); ++i) {
+        const HGCRecHit& hgrh = hits[i];
+        DetId detid = hgrh.detid();
+        unsigned int layerOnSide = (rhtools_.getLayerWithOffset(detid) - 1);
+
+        // set sigmaNoise default value 1 to use kappa value directly in case of
+        // sensor-independent thresholds
+        int thickness_index = rhtools_.getSiThickIndex(detid);
+        if (thickness_index == -1) {
+          thickness_index = maxNumberOfThickIndices_;
+        }
+        double storedThreshold = thresholds_[layerOnSide][thickness_index];
+        if (hgrh.energy() < storedThreshold)
+          continue;  // this sets the ZS threshold at ecut times the sigma noise
+        index++;
+      }
+
+      // Allocate Host SoA will contain one entry for each RecHit above threshold
+      HGCalSoARecHitsHostCollection cells(index, iEvent.queue());
+      auto cellsView = cells.view();
+
+      // loop over all hits and create the Hexel structure, skip energies below ecut
+      // for each layer and wafer calculate the thresholds (sigmaNoise and energy)
+      // once
+      index = 0;
+      for (unsigned int i = 0; i < hits.size(); ++i) {
```

Comment:
could be range based for loop? (also below)

Reply:
the value of the index `i` is needed on [line 117](https://github.com/cms-sw/cmssw/pull/45178/files#diff-2e9e3b4ac8c6bf7965b36258c053f74830945377ed650e41f5e9ab3e1d936a35R117)